### PR TITLE
Add support for Habana DLC

### PR DESCRIPTION
**Pull request by tejaschumbalkar**
*GitHub Issue #, if available:*

Note: If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

### Description
- Add support for Habana DLC
### Tests run
**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [ ] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

## Label Checklist
- [ ] I have added the project label for this PR (*<project_name>* or "Improvement")

## PR Checklist
- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron"/"graviton">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by SergTogul — DISMISSED** [review]

File: test/dlc_tests/conftest.py
Comment (arjkesh) [inline test/dlc_tests/conftest.py:688]: nit: these variables can be plural, i.e.`job_types` and `device_types`

File: test/dlc_tests/conftest.py
Comment (arjkesh) [inline test/dlc_tests/conftest.py:690]: So just for my understanding, this currently makes the assumption that we have 3 different types of image parameterization fixtures:
1) the end of the repo name (why not just exact match on the repo name?)
2) exact match in one of the job types defined here, and job type present in image
3) exact match in one of the device types defined here, and device type present in image

File: test/dlc_tests/conftest.py
Comment (arjkesh) [inline test/dlc_tests/conftest.py:58]: if we are going to make the lookup change, let's add comments grouping these fixtures into the three accepted conditions I mentioned (repo_name fixtures, processor fixtures, job type fixtures)

File: test/dlc_tests/conftest.py
Comment (arjkesh) [inline test/dlc_tests/conftest.py:691]: I think this condition can be fragile and is too linked to assumptions about what the fixtures can be. Let's say we want to add a lookup called "pytorch" - which runs on all pytorch images. Example would be torchaudio sanity test. We'd need to add fixture and also update this condition.

I'm a little concerned about this change because it has the ability to affect all the testing logic under the hood - and I don't want to get into a situation where we aren't running the tests we expect to be running. Have you verified on all test builds that tests are being run as expected? 

I have less context on things like EIA, which I assume could potentially also rely on `<framework>_<job_type>` fixture.

If we make this change, let's also formalize the assumptions in our README
- example: for new processors outside of cpu/gpu/eia/neuron/hpu, please name your repo as `<framework>-<job_type>-<processor>`

File: test/dlc_tests/ecs/pytorch/training/test_ecs_pytorch_training.py
Comment (arjkesh) [inline test/dlc_tests/ecs/pytorch/training/test_ecs_pytorch_training.py:174]: If these fixtures spin up resources, this test will also spin up and down resources each execution. I recommend looking for similar tests that do this and commenting out the fixtures for now, since it could lead to unnecessary resource creation/deletion

File: test/dlc_tests/ec2/tensorflow/training/test_tensorflow_training.py
Comment (arjkesh) [inline test/dlc_tests/ec2/tensorflow/training/test_tensorflow_training.py:271]: are these commented out parameterizations needed?

File: src/main.py
Comment (tejaschumbalkar) [reply src/main.py]: Ran black on all the files on the PR

File: test/dlc_tests/benchmark/ec2/tensorflow/training/test_performance_tensorflow_training.py
Comment (tejaschumbalkar) [reply test/dlc_tests/benchmark/ec2/tensorflow/training/test_performance_tensorflow_training.py]: ack

File: test/dlc_tests/conftest.py
Comment (tejaschumbalkar) [reply test/dlc_tests/conftest.py:368]: This functionality is added to use existing EC2 instance to be used for running test independent of instance type

File: test/dlc_tests/ecs/pytorch/training/test_ecs_pytorch_training.py
Comment (tejaschumbalkar) [reply test/dlc_tests/ecs/pytorch/training/test_ecs_pytorch_training.py]: Removed the fixtures and the placeholder test for ECS/EKS/SageMaker. We will add it back once we have the test content.

File: test/dlc_tests/ec2/tensorflow/training/test_tensorflow_training.py
Comment (tejaschumbalkar) [reply test/dlc_tests/ec2/tensorflow/training/test_tensorflow_training.py]: removed the comments

File: test/dlc_tests/ecs/pytorch/training/test_ecs_pytorch_training.py
Comment (tejaschumbalkar) [reply test/dlc_tests/ecs/pytorch/training/test_ecs_pytorch_training.py]: Removed the fixtures and the placeholder test for ECS/EKS/SageMaker. We will add it back once we have the test content.

File: test/dlc_tests/conftest.py
Comment (tejaschumbalkar) [reply test/dlc_tests/conftest.py:690]: The conditions are correct. `endswith()` is used to replicate the existing logic.

File: test/dlc_tests/conftest.py
Comment (tejaschumbalkar) [reply test/dlc_tests/conftest.py:691]: For the first part, yes we need to add condition that matched the framework name. 
For the second part, the repo names were already in the format of `<framework>-<job_type>-<processor>`. We just were not using it, the way we had is `<framework>-<job_type>` and a fixture like `<processor>_only` which is unnecessary. This logic fixes  it.

File: test/dlc_tests/conftest.py
Comment (arjkesh) [reply test/dlc_tests/conftest.py:368]: that's great - we should consider using this for p3dn/p4d tests

**Review by arjkesh — APPROVED** [review]

**Comment by arjkesh** [comment]
@tejaschumbalkar it looks like there are a few tests failing related to these changes, including quick checks, which will block merge. Let me know if you need help resolving, will remove approval for now, although the changes related to lookup seem good

**Review by arjkesh — CHANGES_REQUESTED** [review]
requesting changes based on test failures

**Comment by tejaschumbalkar** [comment]
Found few issues in neuron/eia. Looking into it.

**Comment by tejaschumbalkar** [comment]
Verified the tests for EC2/EKS/ECS/Sanity are covered for neuron/eia images and the test are not skipped due to the pytest parameter modification.

**Comment by tejaschumbalkar** [comment]
Test are working as expected.

**Review by aws-vrnatham — APPROVED** [review]
Looked at the neuron_only changes and looks good to me

File: test/dlc_tests/conftest.py
Comment (arjkesh) [inline test/dlc_tests/conftest.py:304]: nit: should this be named is_neuron_fixture?

**Review by arjkesh — APPROVED** [review]

File: test/dlc_tests/conftest.py
Comment (tejaschumbalkar) [reply test/dlc_tests/conftest.py:304]: will do a fast follow in a separate PR

**Pull request merged**
